### ISSUE TITLE
Target instance 'builder' pattern support.

### ISSUE THREAD

**Issue by jvmlet** [xebia-france/selma#72]
Hello
My target class instance and all inner fields are instantiated using builder pattern (this is protobuf generated classes).
How does selma support this ?
Thanks

**Comment by slemesle** [comment]
Hello,
I see multiple kinds of builder pattern a code sample would help. But I can say that currently Selma have no support for builder class like that: 

```java
public class User {
	private final String firstName; // required
	private final String lastName; // required
	private final int age; // optional
	private final String phone; // optional
	private final String address; // optional

	private User(UserBuilder builder) {
		this.firstName = builder.firstName;
		this.lastName = builder.lastName;
		this.age = builder.age;
		this.phone = builder.phone;
		this.address = builder.address;
	}

	public String getFirstName() {
		return firstName;
	}

	public String getLastName() {
		return lastName;
	}

	public int getAge() {
		return age;
	}

	public String getPhone() {
		return phone;
	}

	public String getAddress() {
		return address;
	}

	public static class UserBuilder {
		private final String firstName;
		private final String lastName;
		private int age;
		private String phone;
		private String address;

		public UserBuilder(String firstName, String lastName) {
			this.firstName = firstName;
			this.lastName = lastName;
		}

		public UserBuilder age(int age) {
			this.age = age;
			return this;
		}

		public UserBuilder phone(String phone) {
			this.phone = phone;
			return this;
		}

		public UserBuilder address(String address) {
			this.address = address;
			return this;
		}

		public User build() {
			return new User(this);
		}

	}
}
```

For now, there is no constructor parameters mapping, the above builder does not respect Java Bean setter syntax. But you can make Selma work on a class builder if it uses setters, an interceptor can be used to call the build method.

**Comment by facboy** [comment]
I have a branch where I started implementing this, but if I'm honest it can make other things harder to do when the output of a single mapper invocation is final and immutable.  Mapping to mutable beans makes some problems like mapping fields that are nested in the source into the 'root' destination bean trivial.

https://github.com/Byhiras/selma/tree/issue72

**Comment by slemesle** [comment]
@facboy Thanks for this work, as you said it is surely not trivial. I need to take some times to analyse your work the problems caused by this kind of mapping. But this is time consuming, I plan to have a deeper look on by the end of April.

**Comment by facboy** [comment]
Yes, once I had it working it made another part of my work harder so I ended up not using it.

**Comment by anjo-swe** [comment]
@facboy @slemesle any update on this issue? I'm interested in using Selma with either Lombok or Immutables

**Comment by facboy** [comment]
Hmm...it's not very elegant, in my case I am mapping from POJO to google AutoValue beans.  What I ended up doing was generating an intermediate bean representation which itself has a toAutoValue() method.  I get Selma to map to the bean, and then call toAutoValue on it.

You could map to the builder that AutoValue generates if you have made it use 'setX' setters.  but this doesn't work very well with nested objects as the builder expects you to inject the immutable AutoValue bean, not the builder.
